src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,181 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    public static class Node<T> {
+        private Node<T> prevNode;
+        private Node<T> nextNode;
+        private T value;
+
+        public Node(Node<T> prevNode, T value, Node<T> nextNode) {
+            this.prevNode = prevNode;
+            this.value = value;
+            this.nextNode = nextNode;
+        }
+
+        public Node(T value) {
+            this.value = value;
+        }
+    }
+
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
     @Override
     public void add(T value) {
+        linkTail(value);
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndexAdd(index);
+        if (index == size) {
+            linkTail(value);
+            return;
+        }
+        if (index == 0) {
+            linkHead(value);
+            return;
+        }
+        Node<T> current = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(current.prevNode, value, current);
+        if (current.prevNode != null) {
+            current.prevNode.nextNode = newNode;
+        }
+        current.prevNode = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            linkTail(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validateIndex(index);
+        if (index == 0) {
+            return head.value;
+        }
+        if (index == size - 1) {
+            return tail.value;
+        }
+        return getNodeByIndex(index).value;
+
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validateIndex(index);
+        Node<T> current = getNodeByIndex(index);
```

Comment:
This is code repeating. You always check the index before getting node by index. What you should do is move the check logic INSIDE the get method.